templates/commands/render.go

```diff
@@ -235,24 +275,237 @@ func (r *Render) realRun(ctx context.Context, rp *runParams) (outErr error) {
+		return err
+	}
+
+	return nil
+}
+
+func executeSpec(ctx context.Context, spec *model.Spec, sp *stepParams) error {
+	for _, step := range spec.Steps {
+		if err := executeOneStep(ctx, step, sp); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
+type stepParams struct {
+	fs          renderFS
+	inputs      map[string]string
+	scratchDir  string
+	stdout      io.Writer
+	templateDir string
+}
+
+func executeOneStep(ctx context.Context, step *model.Step, sp *stepParams) error {
+	switch {
+	case step.Print != nil:
+		return actionPrint(ctx, step.Print, sp)
+	case step.Include != nil:
+		return actionInclude(ctx, step.Include, sp)
+	case step.RegexReplace != nil:
+		return actionRegexReplace(ctx, step.RegexReplace, sp)
+	case step.StringReplace != nil:
+		return actionStringReplace(ctx, step.StringReplace, sp)
+	case step.GoTemplate != nil:
+		return actionGoTemplate(ctx, step.GoTemplate, sp)
+	default:
+		return fmt.Errorf("internal error: unknown step action type %q", step.Action.Val)
+	}
+}
+
+func errWithPos(pos *model.ConfigPos, fmtStr string, args ...any) error {
+	err := fmt.Errorf(fmtStr, args...)
+	lineStr := fmt.Sprintf("line %d", pos.Line)
+	if pos.Line == 0 {
+		lineStr = "(unknown line number)"
+	}
+	return fmt.Errorf("failed executing template spec file at %s: %w", lineStr, err)
+}
+
+func parseGoTmpl(tpl string) (*template.Template, error) {
+	return template.New("").Option("missingkey=error").Parse(tpl)
+}
+
+func parseAndExecuteGoTmpl(m model.String, inputs map[string]string) (string, error) {
+	goTmpl, err := parseGoTmpl(m.Val)
+	if err != nil {
+		return "", errWithPos(m.Pos, `error compiling as go-template: %w`, err)
+	}
+
+	sb := strings.Builder{}
+	if err := goTmpl.Execute(&sb, inputs); err != nil {
+		return "", errWithPos(m.Pos, "template.Execute() failed: %w", err)
+	}
+
+	return sb.String(), nil
+}
+
+func actionPrint(ctx context.Context, p *model.Print, sp *stepParams) error {
+	msg, err := parseAndExecuteGoTmpl(p.Message, sp.inputs)
+	if err != nil {
+		return err
+	}
+	if !strings.HasSuffix(msg, "\n") {
+		msg += "\n"
+	}
+
+	// We can ignore the int returned from Write() because the docs promise that
+	// short writes always return error.
+	if _, err := sp.stdout.Write([]byte(msg)); err != nil {
+		return fmt.Errorf("error writing to stdout: %w", err)
+	}
+
+	return nil
+}
+
+func actionInclude(ctx context.Context, i *model.Include, sp *stepParams) error {
+	for _, p := range i.Paths {
+		// Paths may contain template expressions, so render them first.
+		walkRelPath, err := parseAndExecuteGoTmpl(p, sp.inputs)
+		if err != nil {
+			return errWithPos(p.Pos, `error compiling as go-template: %w`, err)
+		}
+
+		if err := safeRelPath(p.Val); err != nil {
+			return errWithPos(p.Pos, "invalid path: %w", err)
+		}
+
+		absSrc := filepath.Join(sp.templateDir, walkRelPath)
+		absDst := filepath.Join(sp.scratchDir, walkRelPath)
+
+		if _, err := sp.fs.Stat(absSrc); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
```

Comment:
What if it's a different error? I think we're missing a different return?

Reply:
I also noticed and fixed this :)